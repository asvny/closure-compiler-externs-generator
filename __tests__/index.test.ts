import {
  createApplyDefaults,
  processLibraries,
  FS,
  Library,
} from '../src/index';
import { Volume, createFsFromVolume, DirectoryJSON } from 'memfs';
import { libraries } from './fixtures/libraries';
import * as path from 'path';

const fixturesDir = path.resolve(__dirname, 'fixtures');

function createVolumeAndFs() {
  const volume = new Volume();
  const fs = (createFsFromVolume(volume) as unknown) as FS;

  return {
    fs,
    volume,
  };
}

function normaliseVolumeSnapshot(directoryJSON: DirectoryJSON): DirectoryJSON {
  const newDirectoryJSON: DirectoryJSON = {};

  for (let filePath in directoryJSON) {
    const content = directoryJSON[filePath];
    if (filePath.startsWith(fixturesDir)) {
      filePath = '/' + path.relative(fixturesDir, filePath);
    }
    newDirectoryJSON[filePath] = content;
  }

  return newDirectoryJSON;
}

function snapshotLibraries(
  libraries: Partial<Library> & { moduleName: string }[],
) {
  const tailoredApplyDefaults = createApplyDefaults(fixturesDir);
  const { volume, fs } = createVolumeAndFs();
  processLibraries(
    path.join(fixturesDir, 'out'),
    libraries.map(tailoredApplyDefaults),
    false,
    fs,
  );
  expect(normaliseVolumeSnapshot(volume.toJSON())).toMatchSnapshot();
}

describe('externs-generator', () => {
  describe('generation', () => {
    it.each([false, true])(
      'produces externs for a given set of modules with debug = %s',
      (debug) => {
        expect.hasAssertions();
        const { volume, fs } = createVolumeAndFs();
        processLibraries(path.join(fixturesDir, 'out'), libraries, debug, fs);
        expect(normaliseVolumeSnapshot(volume.toJSON())).toMatchSnapshot();
      },
    );

    it('for scoped modules', () => {
      expect.hasAssertions();
      snapshotLibraries([
        { moduleName: '@scoped/exports-sugar-esm' },
        { moduleName: 'cjs-named-exports' },
        { moduleName: 'main-implicit' },
        { moduleName: 'untyped-cjs-and-esm' },
        { moduleName: 'untyped-cjs' },
      ]);
    });

    it('for various single-export modules', () => {
      expect.hasAssertions();
      snapshotLibraries([
        { moduleName: 'cjs-named-exports' },
        { moduleName: 'main-implicit' },
        { moduleName: 'typings-synonym' },
        { moduleName: 'untyped-cjs-and-esm' },
        { moduleName: 'untyped-cjs' },
      ]);
    });
  });
});
