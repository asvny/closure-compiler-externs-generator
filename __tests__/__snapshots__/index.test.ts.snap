// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`externs-generator generation for scoped modules 1`] = `
Object {
  "/out/cjs_named_exports.js": "// @formatter:off
const __cjs_named_exports = {};
__cjs_named_exports.bar;
__cjs_named_exports.foo;
",
  "/out/main_implicit.js": "// @formatter:off
var foo;
const __main_implicit = {};
",
  "/out/scoped__exports_sugar_esm.js": "// @formatter:off
const __scoped__exports_sugar_esm = {};
__scoped__exports_sugar_esm.barfoo;
__scoped__exports_sugar_esm.foobar;
",
  "/out/untyped_cjs.js": "// @formatter:off
var UntypedCjs;
const __untyped_cjs = {};
__untyped_cjs.useFoo;
",
  "/out/untyped_cjs_and_esm.js": "// @formatter:off
var UntypedCjsAndEsm;
const __untyped_cjs_and_esm = {};
__untyped_cjs_and_esm.useFoo;
",
}
`;

exports[`externs-generator generation for various single-export modules 1`] = `
Object {
  "/out/cjs_named_exports.js": "// @formatter:off
const __cjs_named_exports = {};
__cjs_named_exports.bar;
__cjs_named_exports.foo;
",
  "/out/main_implicit.js": "// @formatter:off
var foo;
const __main_implicit = {};
",
  "/out/typings_synonym.js": "// @formatter:off
const __typings_synonym = {};
__typings_synonym.bar;
__typings_synonym.foo;
",
  "/out/untyped_cjs.js": "// @formatter:off
var UntypedCjs;
const __untyped_cjs = {};
__untyped_cjs.useFoo;
",
  "/out/untyped_cjs_and_esm.js": "// @formatter:off
var UntypedCjsAndEsm;
const __untyped_cjs_and_esm = {};
__untyped_cjs_and_esm.useFoo;
",
}
`;

exports[`externs-generator generation produces externs for a given set of modules with debug = false 1`] = `
Object {
  "/out/main_implicit.js": "// @formatter:off
var foo;
const __main_implicit = {};
",
  "/out/untyped_cjs.js": "// @formatter:off
var UntypedCjs;
const __untyped_cjs = {};
__untyped_cjs.useFoo;
",
}
`;

exports[`externs-generator generation produces externs for a given set of modules with debug = true 1`] = `
Object {
  "/out/main_implicit.js": "// @formatter:off
var foo; // ../node_modules/main-implicit/index.d.ts (2:18)
const __main_implicit = {};
",
  "/out/untyped_cjs.js": "// @formatter:off
var UntypedCjs; // ../node_modules/@types/untyped-cjs/index.d.ts (3:19)
const __untyped_cjs = {};
__untyped_cjs.useFoo; // ../node_modules/@types/untyped-cjs/index.d.ts (4:14)
",
}
`;
